crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -423,6 +424,411 @@ pub fn derive_system_param(input: TokenStream) -> TokenStream {
+                    #(#phantom_field_idents: Default::default(),)*
+                }
+            }
+
+            #[inline]
+            unsafe fn archetype_fetch(&mut self, _archetype_index: usize) -> Self::Item {
+                use #path::query::FilterFetch;
+                #struct_name {
+                    #(#non_filter_field_idents: self.#non_filter_field_idents.archetype_fetch(_archetype_index),)*
+                    #(#filter_field_idents: self.#filter_field_idents.archetype_filter_fetch(_archetype_index),)*
+                    #(#phantom_field_idents: Default::default(),)*
+                }
+            }
+        }
+
+        // SAFETY: update_component_access and update_archetype_component_access are called for each item in the struct
+        unsafe impl #impl_generics #path::query::FetchState for #state_struct_name #ty_generics #where_clause {
+            fn init(world: &mut #path::world::World) -> Self {
+                #state_struct_name {
+                    #(#field_idents: <#query_types as #path::query::WorldQuery>::State::init(world),)*
+                }
+            }
+
+            fn update_component_access(&self, _access: &mut #path::query::FilteredAccess<#path::component::ComponentId>) {
+                #(self.#field_idents.update_component_access(_access);)*
+            }
+
+            fn update_archetype_component_access(&self, _archetype: &#path::archetype::Archetype, _access: &mut #path::query::Access<#path::archetype::ArchetypeComponentId>) {
+                #(self.#field_idents.update_archetype_component_access(_archetype, _access);)*
+            }
+
+            fn matches_archetype(&self, _archetype: &#path::archetype::Archetype) -> bool {
+                true #(&& self.#field_idents.matches_archetype(_archetype))*
+            }
+
+            fn matches_table(&self, _table: &#path::storage::Table) -> bool {
+                true #(&& self.#field_idents.matches_table(_table))*
+            }
+        }
+
+        impl #impl_generics #path::query::WorldQuery for #struct_name #ty_generics #where_clause {
+            type Fetch = #fetch_struct_name #ty_generics;
+            type State = #state_struct_name #ty_generics;
+        }
+
+        #read_only_impl
+    });
+    tokens
+}
+
+/// Implement `FilterFetch` to use a struct as a filter parameter in a query
+#[proc_macro_derive(FilterFetch)]
+pub fn derive_fetch_filter(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let FetchImplTokens {
+        struct_name,
+        fetch_struct_name,
+        state_struct_name,
+        fetch_trait_punctuated_lifetimes,
+        impl_generics,
+        ty_generics,
+        where_clause,
+        struct_has_world_lt,
+        world_lt,
+        state_lt,
+    } = fetch_impl_tokens(&ast);
+
+    // Fetch's HRTBs require this hack to make the implementation compile. I don't fully understand
```

Comment:
```suggestion
    // FilterFetch's HRTBs require this hack to make the implementation compile. I don't fully understand
```

Reply:
This is still `Fetch`. :) I don't implement `FilterFetch` in the macro, it's automatically implemented by `impl<T> FilterFetch for T`. I'm now thinking that macro naming might be confusing, but it seems to be the best way to separate "`FilterFetch`" implementation from `Fetch`.

I believe you saw my previous PoC where it was just `WorldQuery` for both. The macro code was awful there and had even more branches. The implementation for filters looks pretty similar, but it has lots of little details that differ.

Reply:
Aha, got it. I'd like to be able to reduce code duplication, but I suspect that's actually the wrong call here and will lead to frustrating coupling.